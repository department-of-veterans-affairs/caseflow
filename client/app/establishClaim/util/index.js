import { MODIFIER_OPTIONS } from '../constants';
import ROUTING_INFORMATION from '../../constants/Routing';
import { enabledSpecialIssues } from '../../constants/SpecialIssueEnabler.js';

/*
 * This function returns a nicely formatted string for the station of jurisdiction
 * For example:
 *   '397' => '397 - ARC'
 *   '311' => '311 - Pittsburgh, PA'
 */
export const formattedStationOfJurisdiction = (
  stationOfJurisdiction,
  regionalOfficeKey,
  regionalOfficeCities,
  special_issues_revamp
) => {
  let suffix;

  enabledSpecialIssues(special_issues_revamp).forEach((issue) => {
    let issueKey = issue.stationOfJurisdiction && issue.stationOfJurisdiction.key;

    // If the assigned stationOfJurisdiction matches a routed special issue, use the
    // routed station's location
    if (issueKey && issueKey === stationOfJurisdiction) {
      suffix = issue.stationOfJurisdiction.location;
    }
  });

  // ARC is a special snowflake and doesn't show the state (DC)
  if (stationOfJurisdiction === '397') {
    suffix = 'ARC';
  }

  // If there is no routed special issue override, use the default city/state
  if (!suffix) {
    suffix = `${regionalOfficeCities[regionalOfficeKey].city}, ${
      regionalOfficeCities[regionalOfficeKey].state}`;
  }

  return `${stationOfJurisdiction} - ${suffix}`;
};

/*
 * This function gets the set of unused modifiers.
 * This removes any modifiers that have already been
 * used in previous EPs.
 */
export const validModifiers = (endProducts) => {
  let modifiers = MODIFIER_OPTIONS;

  let modifierHash = endProducts.reduce((modifierObject, endProduct) => {
    modifierObject[endProduct.end_product_type_code] = true;

    return modifierObject;
  }, {});

  return modifiers.filter((modifier) => !modifierHash[modifier]);
};

const getRegionalOfficeString = (regionalOfficeKey, regionalOfficeCities) => {
  if (!regionalOfficeKey) {
    return null;
  }

  return `${regionalOfficeKey} - ${
    regionalOfficeCities[regionalOfficeKey].city}, ${
    regionalOfficeCities[regionalOfficeKey].state}`;
};

const getEmailFromConstant = (constant, regionalOfficeKey) => {
  return ROUTING_INFORMATION.codeToEmailMapper[constant[regionalOfficeKey]];
};

export const getSpecialIssuesRegionalOfficeCode = (specialIssuesRegionalOffice, regionalOfficeKey) => {
  switch (specialIssuesRegionalOffice) {
  case 'PMC':
    return ROUTING_INFORMATION.PMC[regionalOfficeKey];
  case 'COWC':
    return ROUTING_INFORMATION.COWC[regionalOfficeKey];
  case 'education':
    return ROUTING_INFORMATION.EDUCATION[regionalOfficeKey];
  default:
    return specialIssuesRegionalOffice;
  }
};

// This method returns a string version of the regional office code. So it takes "R081"
// and returns a string "RO81 - Philadelphia Pension Center, PA"
export const getSpecialIssuesRegionalOffice =
  (specialIssuesRegionalOffice, regionalOfficeKey, regionalOfficeCities) => {
    return getRegionalOfficeString(
      getSpecialIssuesRegionalOfficeCode(specialIssuesRegionalOffice, regionalOfficeKey),
      regionalOfficeCities
    );
  };

export const getSpecialIssuesEmail = (specialIssuesEmail, regionalOfficeKey) => {
  switch (specialIssuesEmail) {
  case 'PMC':
    return getEmailFromConstant(ROUTING_INFORMATION.PMC, regionalOfficeKey);
  case 'COWC':
    return getEmailFromConstant(ROUTING_INFORMATION.COWC, regionalOfficeKey);
  case 'education':
    return getEmailFromConstant(ROUTING_INFORMATION.EDUCATION, regionalOfficeKey);
  default:
    return specialIssuesEmail;
  }
};
