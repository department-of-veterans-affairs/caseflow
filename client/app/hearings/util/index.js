import _ from 'lodash';

export const getReaderLink = (appealVacolsId) => `/reader/appeal/${appealVacolsId}/documents`;

export const orderTheDocket = (docket) =>
  _.orderBy(docket, ['date', 'veteran_last_name', 'appellant_last_name'], ['asc', 'asc', 'asc']);
