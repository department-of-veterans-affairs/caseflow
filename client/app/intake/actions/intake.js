import { ACTIONS, ENDPOINT_NAMES } from '../constants';
import ApiUtil from '../../util/ApiUtil';

const analytics = true;

export const startNewIntake = () => ({
  type: ACTIONS.START_NEW_INTAKE,
  meta: { analytics }
});

export const setFileNumberSearch = (fileNumber) => ({
  type: ACTIONS.SET_FILE_NUMBER_SEARCH,
  payload: {
    fileNumber
  }
});

export const doFileNumberSearch = (formType, fileNumberSearch) => (dispatch) => {
  dispatch({
    type: ACTIONS.FILE_NUMBER_SEARCH_START,
    meta: { analytics }
  });

  const data = {
    file_number: fileNumberSearch,
    form_type: formType
  };

  return ApiUtil.post('/intake', { data }, ENDPOINT_NAMES.START_INTAKE).
    then(
      (response) => {
        dispatch({
          type: ACTIONS.FILE_NUMBER_SEARCH_SUCCEED,
          payload: {
            intake: response.body
          },
          meta: { analytics }
        });
      },
      (error) => {
        const responseObject = error.response.body || {};
        const errorCode = responseObject.error_code || 'default';

        dispatch({
          type: ACTIONS.FILE_NUMBER_SEARCH_FAIL,
          payload: {
            errorCode,
            errorData: responseObject.error_data || {}
          },
          meta: {
            analytics: {
              label: errorCode
            }
          }
        });

        throw error;
      }
    ).
    catch((error) => error);
};

export const setFormType = (formType) => ({
  type: ACTIONS.SET_FORM_TYPE,
  payload: {
    formType
  },
  meta: {
    analytics: {
      label: formType
    }
  }
});

export const clearSearchErrors = () => ({
  type: ACTIONS.CLEAR_SEARCH_ERRORS,
  meta: { analytics }
});

export const setReceiptDate = (receiptDate) => ({
  type: ACTIONS.SET_RECEIPT_DATE,
  payload: {
    receiptDate
  }
});

export const setOptionSelected = (optionSelected) => ({
  type: ACTIONS.SET_OPTION_SELECTED,
  payload: {
    optionSelected
  },
  meta: {
    analytics: {
      label: optionSelected
    }
  }
});

export const toggleCancelModal = () => ({
  type: ACTIONS.TOGGLE_CANCEL_MODAL,
  meta: {
    analytics: {
      label: (nextState) => nextState.intake.cancelModalVisible ? 'show' : 'hide'
    }
  }
});

export const submitCancel = (data) => (dispatch) => {
  dispatch({
    type: ACTIONS.CANCEL_INTAKE_START,
    meta: { analytics }
  });

  return ApiUtil.delete(`/intake/${data.id}`, { data }, ENDPOINT_NAMES.CANCEL_INTAKE).
    then(
      () => dispatch({
        type: ACTIONS.CANCEL_INTAKE_SUCCEED,
        meta: { analytics }
      }),
      (error) => {
        dispatch({
          type: ACTIONS.CANCEL_INTAKE_FAIL,
          meta: { analytics }
        });
        throw error;
      }
    ).
    catch((error) => error);
};

export const submitIntakeCompleteRequest = (intakeId, data) => (dispatch) => {
  return ApiUtil.patch(`/intake/${intakeId}/complete`, data, ENDPOINT_NAMES.COMPLETE_INTAKE).
    then(
      (response) => {
        dispatch({
          type: ACTIONS.COMPLETE_INTAKE_SUCCEED,
          payload: {
            intake: response.body
          },
          meta: { analytics }
        });

        return true;
      },
      (error) => {
        const responseObject = error.response.body || {};
        const responseErrorCode = responseObject.error_code;
        const responseErrorData = responseObject.error_data;
        const responseErrorUUID = responseObject.error_uuid;

        dispatch({
          type: ACTIONS.COMPLETE_INTAKE_FAIL,
          payload: {
            responseErrorCode,
            responseErrorData,
            responseErrorUUID
          },
          meta: { analytics }
        });
        throw error;
      }
    );
};

export const submitIntakeReviewRequest = (intakeId, data) => (dispatch) => {
  return ApiUtil.patch(`/intake/${intakeId}/review`, data, ENDPOINT_NAMES.REVIEW_INTAKE).
    then(
      () => dispatch({
        type: ACTIONS.SUBMIT_REVIEW_SUCCEED,
        meta: { analytics }
      }),
      (error) => {
        const responseObject = error.response.body || {};
        const responseErrorCodes = responseObject.error_codes;

        dispatch({
          type: ACTIONS.SUBMIT_REVIEW_FAIL,
          payload: {
            responseErrorCodes
          },
          meta: {
            analytics: analyticsCallback
          }
        });

        throw error;
      }
    );
};

export const analyticsCallback = (triggerEvent, category, actionName) => {
  triggerEvent(category, actionName, 'any-error');

  _.forEach(
    responseErrorCodes,
    (errorVal, errorKey) => triggerEvent(category, actionName, `${errorKey}-${errorVal}`)
  );
};
