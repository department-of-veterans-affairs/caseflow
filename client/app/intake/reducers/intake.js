import { ACTIONS, REQUEST_STATE } from '../constants';
import { update } from '../../util/ReducerUtil';
import { formatDateStr } from '../../util/DateUtil';
import _ from 'lodash';

const updateFromServerIntake = (state, serverIntake) => {
  return update(state, {
    id: {
      $set: serverIntake.id
    },
    formType: {
      $set: serverIntake.form_type
    },
    asyncJobUrl: {
      $set: serverIntake.async_job_url
    },
    detailEditUrl: {
      $set: serverIntake.detail_edit_url
    },
    unreadMessages: {
      $set: serverIntake.unread_messages
    },
    veteran: {
      name: {
        $set: serverIntake.veteran_name
      },
      formName: {
        $set: serverIntake.veteran_form_name
      },
      fileNumber: {
        $set: serverIntake.veteran_file_number
      },
      isDeceased: {
        $set: serverIntake.veteran_is_deceased
      }
    }
  });
};

export const mapDataToInitialIntake = (data = { serverIntake: {} }) => (
  updateFromServerIntake({
    id: null,
    asyncJobUrl: null,
    detailEditUrl: null,
    formType: null,
    fileNumberSearch: '',
    searchErrorCode: null,
    searchErrorData: {
      duplicateReceiptDate: null,
      duplicateProcessedBy: null,
      veteranMissingFields: null,
      veteranAddressTooLong: null,
      veteranAddressInvalidFields: null,
      veteranCityInvalidFields: null,
      pids: null
    },
    cancelModalVisible: false,
    unreadMessages: false,
    veteran: {
      name: '',
      formName: '',
      fileNumber: '',
      isDeceased: null
    },
    requestStatus: {
      fileNumberSearch: REQUEST_STATE.NOT_STARTED,
      cancel: REQUEST_STATE.NOT_STARTED
    }
  }, data.serverIntake)
);

const resetIntake = () => mapDataToInitialIntake({ serverIntake: {} });

export const intakeReducer = (state = mapDataToInitialIntake(), action) => {
  switch (action.type) {
  case ACTIONS.START_NEW_INTAKE:
    return resetIntake();
  case ACTIONS.SET_FILE_NUMBER_SEARCH:
    return update(state, {
      fileNumberSearch: {
        $set: action.payload.fileNumber
      }
    });
  case ACTIONS.SET_FORM_TYPE:
    return update(state, {
      formType: {
        $set: action.payload.formType
      }
    });
  case ACTIONS.FILE_NUMBER_SEARCH_START:
    return update(state, {
      requestStatus: {
        fileNumberSearch: {
          $set: REQUEST_STATE.IN_PROGRESS
        }
      }
    });
  case ACTIONS.FILE_NUMBER_SEARCH_SUCCEED:
    return updateFromServerIntake(update(state, {
      requestStatus: {
        fileNumberSearch: {
          $set: REQUEST_STATE.SUCCEEDED
        }
      }
    }), action.payload.intake);
  case ACTIONS.FILE_NUMBER_SEARCH_FAIL:
    return update(state, {
      searchErrorCode: {
        $set: action.payload.errorCode
      },
      searchErrorData: {
        duplicateReceiptDate: {
          $set: formatDateStr(action.payload.errorData.receipt_date)
        },
        duplicateProcessedBy: {
          $set: action.payload.errorData.processed_by
        },
        veteranMissingFields: {
          $set: _.join(action.payload.errorData.veteran_missing_fields, ', ')
        },
        veteranAddressTooLong: {
          $set: action.payload.errorData.veteran_address_too_long
        },
        veteranAddressInvalidFields: {
          $set: action.payload.errorData.veteran_address_invalid_fields
        },
        veteranCityInvalidFields: {
          $set: action.payload.errorData.veteran_city_invalid_fields
        },
        veteranCityTooLong: {
          $set: action.payload.errorData.veteran_city_too_long
        },
        veteranDateOfBirthInvalid: {
          $set: action.payload.errorData.veteran_date_of_birth_invalid
        },
        veteranNameSuffixInvalid: {
          $set: action.payload.errorData.veteran_name_suffix_invalid
        },
        pids: {
          $set: _.join(action.payload.errorData.pids, ', ')
        }
      },
      requestStatus: {
        fileNumberSearch: {
          $set: REQUEST_STATE.FAILED
        }
      }
    });
  case ACTIONS.CLEAR_SEARCH_ERRORS:
    return update(state, {
      searchErrorCode: {
        $set: null
      },
      searchErrorData: {
        duplicateReceiptDate: {
          $set: null
        },
        duplicateProcessedBy: {
          $set: null
        },
        veteranMissingFields: {
          $set: null
        },
        veteranAddressTooLong: {
          $set: null
        },
        veteranAddressInvalidFields: {
          $set: null
        },
        veteranCityInvalidFields: {
          $set: null
        },
        pids: {
          $set: null
        }
      }
    });
  case ACTIONS.CANCEL_INTAKE_START:
    return update(state, {
      requestStatus: {
        cancel: {
          $set: REQUEST_STATE.IN_PROGRESS
        }
      }
    });
  case ACTIONS.CANCEL_INTAKE_FAIL:
    return update(state, {
      requestStatus: {
        cancel: {
          $set: REQUEST_STATE.FAILED
        }
      },
      $toggle: ['cancelModalVisible']
    });
  case ACTIONS.CANCEL_INTAKE_SUCCEED:
    return update(resetIntake(), {
      requestStatus: {
        cancel: {
          $set: REQUEST_STATE.SUCCEEDED
        }
      }
    });
  case ACTIONS.TOGGLE_CANCEL_MODAL:
    return update(state, {
      $toggle: ['cancelModalVisible']
    });
  default:
    return state;
  }
};
