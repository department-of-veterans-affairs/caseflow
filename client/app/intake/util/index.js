import _ from 'lodash';
import { REVIEW_OPTIONS, CLAIMANT_ERRORS } from '../constants';
import DATES from '../../../constants/DATES';
import { formatDateStr } from '../../util/DateUtil';

export const getBlankOptionError = (responseErrorCodes, field) => (
  (_.get(responseErrorCodes[field], 0) === 'blank') && 'Please select an option.'
);

export const getClaimantError = (responseErrorCodes) => {
  const errorCode = _.get(responseErrorCodes.claimant, 0);

  return CLAIMANT_ERRORS[errorCode];
};

export const getPageError = (responseErrorCodes) => (
  (_.get(responseErrorCodes.other, 0) === 'unknown_error') && 'Unknown error.'
);

// use this conversion to change between rails model and react radio input
// otherwise we send over a string true/false and reloading turns it into a boolean
// which messes up logic on backend which expects a string
export const convertStringToBoolean = (string) => {
  if (string === 'true') {
    return true;
  } else if (string === 'false') {
    return false;
  }

  return null;
};

export const benefitTypeProcessedInVBMS = (benefitType) => {
  return (benefitType === 'compensation' || benefitType === 'pension');
};

export const isCorrection = (isRating, intakeData) => {
  const isRatingCorrection = isRating && intakeData.hasClearedRatingEp;
  const isNonratingCorrection = !isRating && intakeData.hasClearedNonratingEp;

  return Boolean(isRatingCorrection || isNonratingCorrection);
};

export const getReceiptDateError = (responseErrorCodes, state) => (
  {
    in_future:
      'Receipt date cannot be in the future.',
    before_ramp: 'Receipt Date cannot be earlier than RAMP start date, 11/01/2017.',
    before_ama: `Receipt Date cannot be prior to ${formatDateStr(DATES.AMA_ACTIVATION)}.`,
    before_ramp_receipt_date: 'Receipt date cannot be earlier than the original ' +
      `RAMP election receipt date of ${state.electionReceiptDate}`
  }[_.get(responseErrorCodes.receipt_date, 0)]
);

export const toggleIneligibleError = (hasInvalidOption, selectedOption) => (
  hasInvalidOption && Boolean(selectedOption === REVIEW_OPTIONS.HIGHER_LEVEL_REVIEW.key ||
    selectedOption === REVIEW_OPTIONS.HIGHER_LEVEL_REVIEW_WITH_HEARING.key)
);

export const formatRelationships = (relationships) => {
  return relationships.map((relationship) => {
    const first = _.capitalize(relationship.first_name);
    const last = _.capitalize(relationship.last_name);
    const type = _.capitalize(relationship.relationship_type);

    return {
      value: relationship.participant_id,
      displayText: `${first} ${last}, ${type}`,
      defaultPayeeCode: relationship.default_payee_code
    };
  });
};

export const getDefaultPayeeCode = (state, claimant) => {
  return _.find(state.relationships, { value: claimant }).defaultPayeeCode;
};

export const formatRadioOptions = (options) => {
  return _.map(options, (value, key) => {
    return { value: key,
      displayText: value };
  });
};

export const formatSearchableDropdownOptions = (options) => {
  return _.map(options, (value, key) => {
    return { value: key,
      label: value };
  });
};

export const prepareReviewData = (intakeData, intakeType) => {
  switch (intakeType) {
  case 'appeal':
    return {
      docket_type: intakeData.docketType,
      receipt_date: intakeData.receiptDate,
      claimant: intakeData.claimant,
      veteran_is_not_claimant: intakeData.veteranIsNotClaimant,
      payee_code: intakeData.payeeCode,
      legacy_opt_in_approved: intakeData.legacyOptInApproved
    };
  case 'supplementalClaim':
    return {
      receipt_date: intakeData.receiptDate,
      benefit_type: intakeData.benefitType,
      claimant: intakeData.claimant,
      veteran_is_not_claimant: intakeData.veteranIsNotClaimant,
      payee_code: intakeData.payeeCode,
      legacy_opt_in_approved: intakeData.legacyOptInApproved
    };
  case 'higherLevelReview':
    return {
      informal_conference: intakeData.informalConference,
      same_office: intakeData.sameOffice,
      benefit_type: intakeData.benefitType,
      receipt_date: intakeData.receiptDate,
      claimant: intakeData.claimant,
      veteran_is_not_claimant: intakeData.veteranIsNotClaimant,
      payee_code: intakeData.payeeCode,
      legacy_opt_in_approved: intakeData.legacyOptInApproved
    };
  default:
    return {
      receipt_date: intakeData.receiptDate,
      claimant: intakeData.claimant,
      veteran_is_not_claimant: intakeData.veteranIsNotClaimant,
      payee_code: intakeData.payeeCode,
      legacy_opt_in_approved: intakeData.legacyOptInApproved
    };
  }
};
