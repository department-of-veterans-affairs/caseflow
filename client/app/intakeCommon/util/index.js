import _ from 'lodash';
import { formatDateStringForApi } from '../../util/DateUtil';

export const formatRatings = (ratings, requestIssues = []) => {
  const result = _.keyBy(_.map(ratings, (rating) => {
    return _.assign(rating,
      { issues: _.keyBy(rating.issues, 'reference_id') }
    );
  }), 'profile_date');

  _.forEach(requestIssues, (requestIssue) => {
    // filter out nil dates (request issues that are not yet rated)
    if (requestIssue.profile_date && requestIssue.reference_id) {
      result[requestIssue.profile_date].issues[requestIssue.reference_id].isSelected = true;
    }
  });

  return result;
};

export const validateDate = (date) => {
  const datePattern = /^(0[1-9]|1[0-2])[/](0[1-9]|[12][0-9]|3[01])[/](19|20)\d\d$/;

  if (datePattern.test(date)) {
    return date;
  }

  return null;
};

export const validNonRatedIssue = (issue) => {
  const unvalidatedDate = issue.decisionDate;
  const decisionDate = validateDate(unvalidatedDate);

  if (!issue.description) {
    return false;
  }
  // If there isn't any nonRated category, return 0
  if (!issue.category) {
    return false;
  }
  // If category is unknown issue category, no decision date is necessary.
  if (issue.category === 'Unknown issue category') {
    return true;
  }
  // If category isn't unknown or there's no valid decisionDate, return 0
  if (!decisionDate) {
    return false;
  }

  // If we've gotten to here, that means we've got all necessary parts for a nonRatedIssue to count
  return true;
};

export const formatIssues = (state) => {
  const ratingData = {
    request_issues:
      _(state.ratings).
        map((rating) => {
          return _.map(rating.issues, (issue) => {
            return _.merge(issue, { profile_date: rating.profile_date });
          });
        }).
        flatten().
        filter('isSelected')
  };

  const nonRatingData = {
    request_issues:
      _(state.nonRatedIssues).
        filter((issue) => {
          return validNonRatedIssue(issue);
        }).
        map((issue) => {
          return {
            decision_text: issue.description,
            issue_category: issue.category,
            decision_date: formatDateStringForApi(issue.decisionDate)
          };
        })
  };

  const data = {
    request_issues: _.concat(ratingData.request_issues.value(), nonRatingData.request_issues.value())
  };

  return data;
};
