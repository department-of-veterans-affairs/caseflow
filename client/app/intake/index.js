import React from 'react';
import { combineReducers } from 'redux';
import IntakeFrame from './IntakeFrame';
import { intakeReducer, mapDataToInitialIntake } from './reducers/intake';
import { rampElectionReducer, mapDataToInitialRampElection } from './reducers/rampElection';
import { rampRefilingReducer, mapDataToInitialRampRefiling } from './reducers/rampRefiling';
import { supplementalClaimReducer, mapDataToInitialSupplementalClaim } from './reducers/supplementalClaim';
import { higherLevelReviewReducer, mapDataToInitialHigherLevelReview } from './reducers/higherLevelReview';
import { appealReducer, mapDataToInitialAppeal } from './reducers/appeal';
import { featureToggleReducer, mapDataToFeatureToggle } from './reducers/featureToggles';
import ReduxBase from '../components/ReduxBase';

export const reducer = combineReducers({
  intake: intakeReducer,
  rampElection: rampElectionReducer,
  rampRefiling: rampRefilingReducer,
  supplementalClaim: supplementalClaimReducer,
  higherLevelReview: higherLevelReviewReducer,
  appeal: appealReducer,
  featureToggles: featureToggleReducer
});

export const generateInitialState = (props) => ({
  intake: mapDataToInitialIntake(props),
  rampElection: mapDataToInitialRampElection(props),
  rampRefiling: mapDataToInitialRampRefiling(props),
  supplementalClaim: mapDataToInitialSupplementalClaim(props),
  higherLevelReview: mapDataToInitialHigherLevelReview(props),
  appeal: mapDataToInitialAppeal(props),
  featureToggles: mapDataToFeatureToggle(props)
});

class Intake extends React.PureComponent {
  componentDidMount() {
    if (window.Raven) {
      window.Raven.caseflowAppName = 'intake';
    }
  }

  render() {
    const initialState = generateInitialState(this.props);

    return (
      <ReduxBase initialState={initialState} reducer={reducer} analyticsMiddlewareArgs={['intake']}>
        <IntakeFrame {...this.props} />
      </ReduxBase>
    );
  }
}

export default Intake;
