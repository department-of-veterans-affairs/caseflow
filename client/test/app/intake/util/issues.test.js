import { formatAddedIssues, getAddIssuesFields } from 'app/intake/util/issues';
import { getAddIssuesFieldsSamples } from '../../../data/intake/intakes';

const addedIssues = [
  {
    id: '2',
    decisionIssueId: null,
    description: 'PTSD denied',
    decisionDate: '2020-05-07',
    ineligibleReason: null,
    ineligibleDueToId: null,
    decisionReviewTitle: 'Appeal',
    contentionText: 'PTSD denied',
    vacolsId: null,
    vacolsSequenceId: null,
    vacolsIssue: null,
    endProductCleared: null,
    endProductCode: null,
    withdrawalDate: null,
    editable: true,
    examRequested: null,
    isUnidentified: null,
    notes: null,
    category: null,
    index: '1',
    isRating: true,
    ratingIssueReferenceId: 'def456',
    ratingDecisionReferenceId: null,
    ratingIssueProfileDate: '2020-04-07T00:00:00.000Z',
    approxDecisionDate: '2020-05-07',
    titleOfActiveReview: null,
    rampClaimId: null,
    verifiedUnidentifiedIssue: null
  },
  {
    id: '1',
    decisionIssueId: null,
    description: 'Military Retired Pay - nonrating description',
    decisionDate: '2020-04-27',
    ineligibleReason: null,
    ineligibleDueToId: null,
    decisionReviewTitle: 'Appeal',
    contentionText: 'Military Retired Pay - nonrating description',
    vacolsId: null,
    vacolsSequenceId: null,
    vacolsIssue: null,
    endProductCleared: null,
    endProductCode: null,
    withdrawalDate: null,
    editable: true,
    examRequested: null,
    isUnidentified: null,
    notes: null,
    category: 'Military Retired Pay',
    isRating: false,
    ratingIssueReferenceId: null,
    ratingDecisionReferenceId: null,
    ratingIssueProfileDate: '1970-01-01T00:00:00.000Z',
    approxDecisionDate: '2020-04-27',
    titleOfActiveReview: null,
    rampClaimId: null,
    verifiedUnidentifiedIssue: null
  }
];

const expected = [
  {
    index: 0,
    id: '2',
    text: 'PTSD denied',
    date: '2020-05-07',
    notes: null,
    titleOfActiveReview: null,
    decisionDate: '2020-05-07T00:00:00.000Z',
    beforeAma: false,
    ineligibleReason: null,
    rampClaimId: null,
    vacolsId: null,
    vacolsSequenceId: null,
    vacolsIssue: null,
    withdrawalDate: null,
    endProductCleared: null,
    editable: true,
    examRequested: null,
    decisionIssueId: null,
    ratingIssueReferenceId: 'def456',
    ratingDecisionReferenceId: null
  },
  {
    index: 1,
    id: '1',
    text: 'Military Retired Pay - nonrating description',
    date: '2020-04-27',
    beforeAma: false,
    ineligibleReason: null,
    vacolsId: null,
    vacolsSequenceId: null,
    vacolsIssue: null,
    decisionReviewTitle: 'Appeal',
    withdrawalDate: null,
    endProductCleared: null,
    category: 'Military Retired Pay',
    editable: true,
    examRequested: null,
    decisionIssueId: null
  }
];

const testVeteran = 'Bob Something (000000001)';

const intakesCases = {
  docketTypeNotHearing: 0,
  hearingTypeNull: 1,
  hearingTypePresent: 2,
};

const hearingTypeIsPresent = (fields) => {
  // findIndex returns -1 if an entry with field = 'hearing type' isn't found.
  let idx = fields.findIndex((entry) => entry.field.toLowerCase() === 'hearing type');

  return idx !== -1;
};

describe('formatAddedIssues', () => {
  describe('nonrating issues', () => {
    describe('adding new', () => {
      // Here we expect it to prepend the category to the text description
      const issues = addedIssues.map((issue) => ({
        ...issue,
        id: null,
        description: issue.description.replace(`${issue.category} - `, '')
      }));

      test('returns correctly formatted issues', () => {
        const res = formatAddedIssues(issues);

        expect(res).toMatchSnapshot();
        expect(res[1]).toMatchObject({ ...expected[1], id: null });
      });
    });

    describe('editing existing', () => {
      test('returns correctly formatted issues', () => {
        // Here the text description already contains category, so we shouldn't be prepending
        const res = formatAddedIssues(addedIssues);

        expect(res).toMatchSnapshot();
        expect(res[1]).toMatchObject(expected[1]);
      });
    });
  });
});

describe('getAddIssueFields - form_type: appeal - Hearing type field', () => {
  it('is not present because docket type is not hearing', async () => {
    const fields = getAddIssuesFields(
      'appeal',
      testVeteran,
      getAddIssuesFieldsSamples[intakesCases.docketTypeNotHearing]
    );

    expect(hearingTypeIsPresent(fields)).toBe(false);
  });

  it('is not present because hearing type is null despite docket type being hearing', async () => {
    const fields = getAddIssuesFields(
      'appeal',
      testVeteran,
      getAddIssuesFieldsSamples[intakesCases.hearingTypeNull]
    );

    expect(hearingTypeIsPresent(fields)).toBe(false);
  });

  it('is present because docket type is hearing and hearing type is not null', async () => {
    const fields = getAddIssuesFields(
      'appeal',
      testVeteran,
      getAddIssuesFieldsSamples[intakesCases.hearingTypePresent]
    );

    expect(hearingTypeIsPresent(fields)).toBe(true);
  });
});

