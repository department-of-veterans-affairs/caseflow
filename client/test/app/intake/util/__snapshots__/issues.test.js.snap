// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatAddedIssues nonrating issues adding new returns correctly formatted issues 1`] = `
Array [
  Object {
    "beforeAma": false,
    "correctionType": undefined,
    "date": "2020-05-07",
    "decisionDate": 2020-05-07T00:00:00.000Z,
    "decisionIssueId": null,
    "editable": true,
    "editedDescription": undefined,
    "eligibleForSocOptIn": undefined,
    "eligibleForSocOptInWithExemption": undefined,
    "endProductCleared": null,
    "endProductCode": null,
    "examRequested": null,
    "id": null,
    "index": 0,
    "ineligibleReason": null,
    "mstChecked": undefined,
    "mst_status_update_reason_notes": undefined,
    "notes": null,
    "pactChecked": undefined,
    "pact_status_update_reason_notes": undefined,
    "rampClaimId": null,
    "ratingDecisionReferenceId": null,
    "ratingIssueReferenceId": "def456",
    "sourceReviewType": undefined,
    "text": "PTSD denied",
    "timely": undefined,
    "titleOfActiveReview": null,
    "untimelyExemption": undefined,
    "untimelyExemptionCovid": undefined,
    "untimelyExemptionNotes": undefined,
    "vacolsId": null,
    "vacolsIssue": null,
    "vacolsSequenceId": null,
    "vbmsMstChecked": undefined,
    "vbmsPactChecked": undefined,
    "withdrawalDate": null,
    "withdrawalPending": undefined,
  },
  Object {
    "beforeAma": false,
    "benefitType": undefined,
    "category": "Military Retired Pay",
    "correctionType": undefined,
    "date": "2020-04-27",
    "decisionIssueId": null,
    "decisionReviewTitle": "Appeal",
    "editable": true,
    "editedDecisionDate": undefined,
    "editedDescription": undefined,
    "eligibleForSocOptIn": undefined,
    "eligibleForSocOptInWithExemption": undefined,
    "endProductCleared": null,
    "endProductCode": null,
    "examRequested": null,
    "id": null,
    "index": 1,
    "ineligibleReason": null,
    "isPreDocketNeeded": undefined,
    "mstChecked": undefined,
    "mst_status_update_reason_notes": undefined,
    "pactChecked": undefined,
    "pact_status_update_reason_notes": undefined,
    "text": "Military Retired Pay - nonrating description",
    "timely": undefined,
    "untimelyExemption": undefined,
    "untimelyExemptionCovid": undefined,
    "untimelyExemptionNotes": undefined,
    "vacolsId": null,
    "vacolsIssue": null,
    "vacolsSequenceId": null,
    "vbmsMstChecked": undefined,
    "vbmsPactChecked": undefined,
    "withdrawalDate": null,
    "withdrawalPending": undefined,
  },
]
`;

exports[`formatAddedIssues nonrating issues editing existing returns correctly formatted issues 1`] = `
Array [
  Object {
    "beforeAma": false,
    "correctionType": undefined,
    "date": "2020-05-07",
    "decisionDate": 2020-05-07T00:00:00.000Z,
    "decisionIssueId": null,
    "editable": true,
    "editedDescription": undefined,
    "eligibleForSocOptIn": undefined,
    "eligibleForSocOptInWithExemption": undefined,
    "endProductCleared": null,
    "endProductCode": null,
    "examRequested": null,
    "id": "2",
    "index": 0,
    "ineligibleReason": null,
    "mstChecked": undefined,
    "mst_status_update_reason_notes": undefined,
    "notes": null,
    "pactChecked": undefined,
    "pact_status_update_reason_notes": undefined,
    "rampClaimId": null,
    "ratingDecisionReferenceId": null,
    "ratingIssueReferenceId": "def456",
    "sourceReviewType": undefined,
    "text": "PTSD denied",
    "timely": undefined,
    "titleOfActiveReview": null,
    "untimelyExemption": undefined,
    "untimelyExemptionCovid": undefined,
    "untimelyExemptionNotes": undefined,
    "vacolsId": null,
    "vacolsIssue": null,
    "vacolsSequenceId": null,
    "vbmsMstChecked": undefined,
    "vbmsPactChecked": undefined,
    "withdrawalDate": null,
    "withdrawalPending": undefined,
  },
  Object {
    "beforeAma": false,
    "benefitType": undefined,
    "category": "Military Retired Pay",
    "correctionType": undefined,
    "date": "2020-04-27",
    "decisionIssueId": null,
    "decisionReviewTitle": "Appeal",
    "editable": true,
    "editedDecisionDate": undefined,
    "editedDescription": undefined,
    "eligibleForSocOptIn": undefined,
    "eligibleForSocOptInWithExemption": undefined,
    "endProductCleared": null,
    "endProductCode": null,
    "examRequested": null,
    "id": "1",
    "index": 1,
    "ineligibleReason": null,
    "isPreDocketNeeded": undefined,
    "mstChecked": undefined,
    "mst_status_update_reason_notes": undefined,
    "pactChecked": undefined,
    "pact_status_update_reason_notes": undefined,
    "text": "Military Retired Pay - nonrating description",
    "timely": undefined,
    "untimelyExemption": undefined,
    "untimelyExemptionCovid": undefined,
    "untimelyExemptionNotes": undefined,
    "vacolsId": null,
    "vacolsIssue": null,
    "vacolsSequenceId": null,
    "vbmsMstChecked": undefined,
    "vbmsPactChecked": undefined,
    "withdrawalDate": null,
    "withdrawalPending": undefined,
  },
]
`;
